### ISSUE TITLE
Add training and fine-tuning of bidirectional LM - following ELMO

### ISSUE THREAD

**Issue by PiotrCzapla** [n-waves/ulmfit-multilingual#4]
- [x] BiLM implementation
  - [x] Saving the weights and loading them later
  - [ ] Handle AR & TAR
  - [x] Using biLM for classification tasks to see how much we can improve over SOTA. The fastest and most meaningful comparison will be if we manage to improve sentiment classification over the original ULMFiT paper. @sebastianruder can you propose a dataset and do you have the best parameters?

**Comment by dust0x** [comment]
[Here](https://github.com/allenai/allennlp/blob/master/allennlp/modules/elmo_lstm.py) looks like a good starting point for this problem. Any recommendations on how we may want to implement this with the existing AWD-LSTM/QRNN encoders?

**Comment by sebastianruder** [comment]
Another good reference probably is Section 3.1 of the [ELMo paper](https://arxiv.org/abs/1802.05365). The bidirectional LM simply minimizes the sum of the forward and the backward losses. The forward and backward hidden states are kept separate and don't interact during LM training. Word embeddings and the softmax layer are shared between LSTMs, though.
We want to keep a separate forward and backward QRNN. For each input sequence, we then also obtain the backward sequence. Both sequences are fed into the corresponding LSTM, we get their respective outputs, and calculate the respective losses.

**Comment by sebastianruder** [comment]
The better file to look at is [this one](https://github.com/allenai/allennlp/blob/master/allennlp/models/bidirectional_lm.py#L65). This is the actual implementation of the bidirectional LM in AllenNLP.

**Comment by PiotrCzapla** [comment]
@dust0x  I'm on it as well. But let's do that together as we would need a good understanding of the ulmfit training code to be able to propose other training schemas, Like:
- multi-gpu training (we might need that for biLM)
- BiLM training where the last fully connected layers use both LMs to make word predictions. (an approach closer to BERT)
- to implement MaskedLM = this will be super slow as we can't reuse the previous hidden states.

**Comment by sebastianruder** [comment]
I've created a branch and an initial PR for this here: https://github.com/n-waves/ulmfit-multilingual/pull/6. The full forward pass and loss computation are not yet implemented. 
As far as I can see, the main thing that we need to do is figure out how we get the reverse input. Do we want to do this in the batcher or in the forward method of the model?
For simplicity, it might make sense to just do this in the batcher as we also need to add the targets of the backward model.

**Comment by PiotrCzapla** [comment]
It seems like we could do it inside of forward as well, given that backward LM is simply reversing the input tensor, 
https://github.com/fastai/fastai/blob/1a9adb6d3abe9a188f39fe0823e02badeb99d3b0/fastai/text/data.py#L236

do you know why RNNTrainer callback (https://github.com/fastai/fastai/blob/d932da02d96444017ef640d94c3a96fd8ea99313/fastai/callbacks/rnn.py) saves raw_output?
We might modify it as well as we are going to return more states, or maybe we concatenate them.

**Comment by dust0x** [comment]
I agree with Piotr, it makes sense to do this inside forward for a very simple reason that it saves us from having to pass an additional argument to the batching function.

A minor issue to think about could be (please correct me if I might have misunderstood something here) that while setting the batch size it isn't immediately obvious if `bs=64` means 32 samples for forward and backward input or 64 of each.

**Comment by sebastianruder** [comment]
Yes, I agree doing it in forward would be cleaner. At the moment, there doesn't seem to be an easy way to reverse a tensor in Pytorch, though. So we'd probably have to do convert the tensor to numpy and back, which might result in a performance hit.
I think saving of the raw_output was done as a convention before, but we should be able to remove this. Yes, I think we have to modify the states that we return.

**Comment by PiotrCzapla** [comment]
I've read through the RNN code in fastai and we have to modify `data loaders` to define the target `y` variable, as we don't have access to it in `forward` function.  (I miss tensorflow here were model_fn gets `x` & `y` and let you define loss, all in one place).  Moreover we need to move the window of size ~bppt through text from left to right  for regular `LM` and from right to left for backward `LM`. Otherwise the hidden states between batches will not match for one LM.
So we need to do reversing of the text in data loader anyway to have the backward LM train properly between batches.

Here is what I've learned so far about the LanguageModelLoader.
It spit outs `(x, y)`, where `x` has a shape of `(seq_len, bs)`, and `y` has `(seq_len * bs)`. `y` is actually reshaped from `(seq_len, bs)` and has almost the same content as `x` but moved one word down.

Conceptually I would simply stack two such data loaders together and output tuples: `((fwd_x, bwd_x), (y))`. Where `y` is a  a stacked `fwd_y` and `bwd_y`   so that it matches [cross_entropy](https://pytorch.org/docs/stable/nn.html#torch.nn.functional.cross_entropy) expected shape which is (N,d1,d2,...,dK) . The only issue with this approach is that seq_len is ranodmly changed so I will have to rewrite the LanguageModelLoader but to make sure that it can be selected in sync between backward and foward data loader.

I'm going to start with writing tests and defining some API. I will start working with your proposal but Instead of creating a BiLM  from coping the code of LMCore I will try to reuse it  so that we can easily put two models to two separate GPU's if needed.

I will create a PR with a draft of the API, and some tests stubs, that way one could write tests and another person can write the code.

Sebastian regarding reversing tensors, you can do this as follows (I've found it when I was looking how the RNN can be go in reverse direction in pytorch):
```
t=torch.LongTensor([10,20,30,40,50]);print(t)
idx=torch.arange(t.shape[0]-1,-1,-1)
print(t.index_select(0,idx))
```

**Comment by PiotrCzapla** [comment]
Here is something I've drawn when I read through the code you migth find it useful 
![paper journal 1](https://user-images.githubusercontent.com/340180/48447643-c0c74780-e79c-11e8-9769-9829158f2590.png)

**Comment by sebastianruder** [comment]
Thanks for writing your thoughts down in such a detailed way!
Yes, I agree it'd be nice if we can reuse the existing LMCore. I thought for simplicity, it might be better to just create a new one, as we otherwise have to add additional functionality to ensure that word embedding and softmax layers are shared between the models and for jointly optimizing the losses. Combining the existing components would definitely be preferable.
Do you prefer to finish the implementation of this? In that case, I can start setting up the evaluation on XNLI.

**Comment by PiotrCzapla** [comment]
I won't be able to help with XNLI until I understand fully how the new fastai implementation works, so it makes sense that you jump right on XNLI. The good thing is that it is almost ready, the training runs on two LM's.

The only think to figure out conceptually is how to update RNNTrainer to use both LMs, currently, it uses the last one. It adds something to the loss denoted as AR & TAR. Do you know  what it stands for?

```
    def on_backward_begin(self, last_loss:Rank0Tensor, last_input:Tensor, **kwargs):
        #Adjusts the lr to the bptt selected
        if self.adjust: self.learn.opt.lr *= last_input.size(0) / self.bptt
        #AR and TAR
        if self.alpha != 0.:  last_loss += (self.alpha * self.out[-1].pow(2).mean()).sum()
        if self.beta != 0.:
            h = self.raw_out[-1]
            if len(h)>1: last_loss += (self.beta * (h[1:] - h[:-1]).pow(2).mean()).sum()
        return last_loss
```

**Comment by sebastianruder** [comment]
Sounds good. 👍 
No idea. Perhaps Sylvain can help? Sylvain, could you tell us what the AR and TAR terms added to the loss function are? @sgugger

**Comment by PiotrCzapla** [comment]
Without tweaking the AR & TAR, we have BiLM that trains quite okey, on wikitext-2
The bits that we are still missing are (I've updated the todos in the first comment): 
> * [x]  BiLM implementation
> * [ ]  Saving the weights and loading them later
> * [ ]  Handle AR & TAR
> * [ ]  Using biLM for classification tasks to see how much we can improve over SOTA. The fastest and most meaningful comparison will be if we manage to improve sentiment classification over the original ULMFiT paper. @sebastianruder can you propose a dataset and do you have the best parameters?


```
$ python -m ulmfit.pretrain_lm data/en/wikitext-2 --qrnn=True --cuda-id=1 --name=wt-2-bi-qrnn --num_epochs=10 --bidir=True --bs 40

Batch size: 40
Max vocab: 60000
Using QRNNs...
Size of vocabulary: 33279
First 10 words in vocab: the, <pad>, ,, ., of, <unk>, and, in, to, <eos>
Saving vocabulary...
epoch  train_loss  valid_loss  accuracy_fwd  accuracy_bwd
1      5.693565    5.505599    0.209215      0.215756
2      4.958893    4.962934    0.237294      0.251355
3      4.609792    4.795688    0.246843      0.261626
4      4.324767    4.729492    0.250422      0.265749
5      4.041284    4.704578    0.255372      0.271331
6      3.702716    4.739176    0.255133      0.268892
7      3.256590    4.849097    0.252094      0.264927
8      2.734393    5.024333    0.247509      0.255769
9      2.274600    5.198886    0.242841      0.247664
10     1.998600    5.279930    0.240350      0.244604
```
time: 26:32.04 total

```
$ python -m ulmfit.pretrain_lm data/en/wikitext-2 --qrnn=False --cuda-id=0 --name=wt-2-lstm --num_epochs=10 --bs 40

Batch size: 40
Max vocab: 60000
Size of vocabulary: 33279
First 10 words in vocab: the, <pad>, ,, ., of, <unk>, and, in, to, <eos>
Saving vocabulary...
epoch  train_loss  valid_loss  accuracy
1      5.639165    5.431065    0.219476
2      4.971909    4.967250    0.244650
3      4.701672    4.794151    0.255182
4      4.447895    4.683326    0.260905
5      4.171962    4.638478    0.265196
6      3.869245    4.629985    0.266243
7      3.503400    4.671610    0.264953
8      3.077750    4.771190    0.261643
9      2.702689    4.871647    0.256242
10     2.486330    4.924218    0.254312
```
time: 15:42.03 total

**Comment by PiotrCzapla** [comment]
Apparently, RNNTrainer has impressive documentation: https://docs.fast.ai/callbacks.rnn.html i need to check it more often, where AR i TAR is as well explained.

**Comment by PiotrCzapla** [comment]
BiLM training on wikitext-103,  bs=64, using 30GB memory, but it has better results than single LM training after the same amount of epochs. I had 3.99 val_loss  after 10 epochs
```
epoch  train_loss  valid_loss  accuracy_fwd  accuracy_bwd    
1      4.141750    4.114931    0.307700      0.320260                                                                    
2      4.078711    4.061186    0.310221      0.320550                                                                     
3      4.108722    4.081286    0.308383      0.316431                                                                     
4      4.054437    4.040497    0.311229      0.319658                                                                     
5      4.014880    3.966010    0.318588      0.327607                                                                     
6      3.926770    3.897305    0.326874      0.333171                                                                     
7      3.855364    3.816452    0.335632      0.342045                                                                     
8      3.761600    3.745798    0.343413      0.350675
```

**Comment by sebastianruder** [comment]
Nice! I'd assume sharing the word embedding and softmax parameters is useful. :)

**Comment by PiotrCzapla** [comment]
We have the first version of bi classifier. The names are awful and I need to clean up where the weights are stored. But at least we can see first results. And they are worst than single classifier unfortunately.

**Comment by PiotrCzapla** [comment]
I'm training BiLSTM and BiQRNN but it seems we have another issue as the training seems to be two times slower than previously "6h" per epoch for qrnn 4 layers, 3:40h per epoch for lstm. :(

There is a chance that my DGX-Station is being used by some other team and it is slower because of that and I don't see it on the vm but I have no way to check as all other GPUs I have are 1080ti. @sebastianruder  can you run a bilstm training on your hardware?

This takes me 6h
`python -m ulmfit lm --dataset-path data/wiki/wikitext-103 --bidir=True --qrnn=True --nl=4 --tokenizer=vf --name 'nl4e10' --cuda-id 1 -  train 10 --drop-mult=0 --bs 32`

This takes me 3.4h
`python -m ulmfit lm --dataset-path data/wiki/wikitext-103 --bidir=True --qrnn=False --nl=3 --tokenizer=vf --name 'nl3e10' --cuda-id 0 -  train 10 --drop-mult=0 --bs 32`

**Comment by PiotrCzapla** [comment]
I've got back the v100, and I'm testing it again and an epoch of  bidir lstm takes 3h, bidir qrnn takes 2h and it is the same code as previously so it was an issue with the v100 I had.  Uff. But let's  see how it works with new fastai and recent ulmfit-multilingual

**Comment by PiotrCzapla** [comment]
I've tested the bidir that uses lstm and `vf60k` on imdb, with BiPoolingLinearClassifier and it performs as good as single LM:
Loss and accuracy using (cls_best): [0.16246703, tensor(0.9476)]

@sebastianruder can you think about other bidirectional heads for classification?

**Comment by sebastianruder** [comment]
Hmm. Can you remind me again: Is the output layer just a linear classifier or do we have an additional hidden layer with batch norm as in the original ULMFiT?

**Comment by PiotrCzapla** [comment]
Same as in original ULMFiT I've simply extended the first layer to consume both LSTMs outputs. But let's ignore it for a time being until we have some results for MLDoc.

**Comment by tpietruszka** [comment]
Just a quick note. Recently I, too, have tried to experiment with different classifier heads, both for the unidirectional language model and concatenating outputs of 2 language models (trained and fine-tuned separately).  The performance seems to be very similar between 1 and 2 directions. Adding attention does not help (maybe slightly). 

Any ideas why it might be worse than training 2 separate classifiers (fwd and bwd) and averaging their results? I was so sure it would be better....

**Comment by PiotrCzapla** [comment]
I didn't have seen improvements either. It is likely that IMDB is a bit too easy for the improvements to show. I'm sure you would see something on XNLI. Closing for the time being as we are not using bidirectional training anymore and it is more likely we get the to implement something similar to ELECTRA that would let us use biqrnns
